### PR TITLE
Implement journald logging driver

### PR DESCRIPTION
**Pull request by fahedouch**
enhance https://github.com/containerd/nerdctl/issues/1039 with `journald` log driver

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -258,9 +258,10 @@ RUN go env GOVERSION > /GOVERSION
 FROM base AS test-integration
 ARG DEBIAN_FRONTEND=noninteractive
 # `expect` package contains `unbuffer(1)`, which is used for emulating TTY for testing
+# libsystemd-dev: for sdjournal
 RUN apt-get update && \
   apt-get install -qq -y \
-  expect git
+  expect git libsystemd-dev
```
Comment (AkihiroSuda) [inline Dockerfile:264]: Not a huge fan of cgo.

If cgo is necessary we should split this plugin into a third party repo and call it via `--log-opt binary://`

Comment (AkihiroSuda) [reply Dockerfile:264]: And what's the license of this?

File: pkg/logging/logging.go [hunk 2]
```diff
@@ -88,53 +73,41 @@ func getLoggerFunc(dataStore string) (logging.LoggerFunc, error) {
...
+// LogDriverConfig is defined per a driver
+type LogDriverConfig struct {
+	Driver string            `json:"driver"`
+	Opts   map[string]string `json:"opts,omitempty"`
+}
+
+// LogConfigFilePath returns the path of log-config.json
+func LogConfigFilePath(dataStore, ns, id string) string {
+	return filepath.Join(dataStore, "containers", ns, id, "log-config.json")
+}
```
Comment (AkihiroSuda) [inline pkg/logging/logging.go:113]: No need to move these funcs

File: pkg/logging/logging.go [hunk 3]
```diff
@@ -23,21 +23,22 @@ import (
...
 
 const (
 	// MagicArgv1 is the magic argv1 for the containerd runtime v2 logging plugin mode.
 	MagicArgv1 = "_NERDCTL_INTERNAL_LOGGING"
 	MaxSize    = "max-size"
 	MaxFile    = "max-file"
+	DriverType = "logging-driver"
 )
 
+type NerdctlLoggingInterface interface {
```
Comment (AkihiroSuda) [inline pkg/logging/logging.go:38]: We can just call this `LogDriver`

File: pkg/logging/logging.go [hunk 4]
```diff
@@ -23,21 +23,22 @@ import (
...
-	units "github.com/docker/go-units"
-	"github.com/natefinch/lumberjack"
 )
 
 const (
 	// MagicArgv1 is the magic argv1 for the containerd runtime v2 logging plugin mode.
 	MagicArgv1 = "_NERDCTL_INTERNAL_LOGGING"
 	MaxSize    = "max-size"
 	MaxFile    = "max-file"
+	DriverType = "logging-driver"
```
Comment (AkihiroSuda) [inline pkg/logging/logging.go:35]: What's this?

File: pkg/logging/journald_logger.go [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"io"
+	"sync"
+
+	"github.com/containerd/containerd/runtime/v2/logging"
+	"github.com/coreos/go-systemd/journal"
```
Comment (AkihiroSuda) [inline pkg/logging/journald_logger.go:27]: Why not use v22?

File: cmd/nerdctl/run_test.go [hunk 6]
```diff
@@ -245,3 +246,45 @@ func TestRunWithLogOpt(t *testing.T) {
...
+	jconf := sdjournal.JournalReaderConfig{
+		Matches: []sdjournal.Match{
+			{
+				Field: sdjournal.SD_JOURNAL_FIELD_SYSLOG_IDENTIFIER,
+				Value: syslogIdentifier,
+			},
+		},
+	}
+
+	jr, err := sdjournal.NewJournalReader(jconf)
```
Comment (AkihiroSuda) [inline cmd/nerdctl/run_test.go:274]: Can we just exec `journalctl` to remove cgo dependency?

File: README.md [hunk 7]
```diff
@@ -473,11 +473,12 @@ Metadata flags:
 - :nerd_face: `--pidfile`: file path to write the task's pid. The CLI syntax conforms to Podman convention.
 
 Logging flags:
-- :whale: `--log-driver=(json-file)`: Logging driver for the container (default `json-file`).
+- :whale: `--log-driver=(json-file|journald)`: Logging driver for the container (default `json-file`).
```
Comment (AkihiroSuda) [inline README.md:476]: Could you update the shell completion too

File: cmd/nerdctl/run_test.go [hunk 8]
```diff
@@ -245,3 +246,45 @@ func TestRunWithLogOpt(t *testing.T) {
...
+	assert.NilError(t, err)
+
+	b := make([]byte, 20)
+	found := 0
+	check := func(log poll.LogT) poll.Result {
+		c, err := jr.Read(b)
+		assert.NilError(t, err)
```
Comment (AkihiroSuda) [inline cmd/nerdctl/run_test.go:281]: Can we verify the content? How is it formatted?

File: Dockerfile (hunk 1 above)
Comment (Zheaoli) [reply Dockerfile:264]: I think we should not think about cgo, I think the dynamic link library may cause some potential problems in some distribution.

File: cmd/nerdctl/run_test.go (hunk 6 above)
Comment (Zheaoli) [reply cmd/nerdctl/run_test.go:274]: I think journalctl is not enough if we want to read logs of what we have written.

I think it would be better to we can design the journal logger as an unreadable logger, and use `JSON` logger  and journal logger at the same time

Comment (AkihiroSuda) [reply cmd/nerdctl/run_test.go:274]: Why journalctl is not enough?
If journalctl is not enough, how is this expected to be used?

Comment (fahedouch) [reply cmd/nerdctl/run_test.go:274]: Journalctl is possible with  SYSLOG_IDENTIFIER   filter. need for cgo when using journactl in tests. 
@Zheaoli I want to avoid making depenciences between drivers

File: README.md (hunk 7 above)
Comment (fahedouch) [reply README.md:476]: where ?

File: Dockerfile (hunk 1 above)
Comment (fahedouch) [reply Dockerfile]: fixed

File: pkg/logging/logging.go (hunk 2 above)
Comment (fahedouch) [reply pkg/logging/logging.go]: fixed

File: pkg/logging/logging.go (hunk 3 above)
Comment (fahedouch) [reply pkg/logging/logging.go]: fixed

File: pkg/logging/logging.go (hunk 4 above)
Comment (fahedouch) [reply pkg/logging/logging.go]: fixed

File: pkg/logging/journald_logger.go (hunk 5 above)
Comment (fahedouch) [reply pkg/logging/journald_logger.go]: fixed

File: README.md (hunk 7 above)
Comment (fahedouch) [reply README.md:476]: fixed

File: cmd/nerdctl/run_test.go [hunk 9]
```diff
@@ -245,3 +246,57 @@ func TestRunWithLogOpt(t *testing.T) {
...
+
+	defer base.Cmd("rm", "-f", containerName).AssertOK()
+	base.Cmd("run", "--network", "none", "-d", "--log-driver", "journald", "--name", containerName, testutil.CommonImage,
+		"sh", "-euxc", "echo foo; echo bar").AssertOK()
+
+	time.Sleep(3 * time.Second)
+	inspectedContainer := base.InspectContainer(containerName)
+	found := 0
+	check := func(log poll.LogT) poll.Result {
+		res := icmd.RunCmd(icmd.Command("journalctl", "--no-pager", fmt.Sprintf("SYSLOG_IDENTIFIER=%s", inspectedContainer.ID[:12])))
```
Comment (AkihiroSuda) [inline cmd/nerdctl/run_test.go:265]: needs something like `--since "1 hour ago"`, otherwise journalctl produces a huge hog

**Comment by AkihiroSuda** [comment]
Can we implement `nerdctl logs` for journald, or at least print a human-readable error to clarify that the feature is not implemented yet?

File: README.md [hunk 10]
```diff
@@ -694,7 +695,9 @@ Usage: `nerdctl logs [OPTIONS] CONTAINER`
 Flags:
 - :whale: `--f, --follow`: Follow log output
 - :whale: `--since`: Show logs since timestamp (e.g. 2013-01-02T13:23:37Z) or relative (e.g. 42m for 42 minutes)
+  - for journald log driver logs since relative should be in this format `42 minutes ago`
```
Comment (AkihiroSuda) [inline README.md:698]: Can we just use the same format?
If difficult this flag can be implemented in a separate PR.

File: cmd/nerdctl/logs.go [hunk 11]
```diff
@@ -76,80 +78,113 @@ func logsAction(cmd *cobra.Command, args []string) error {
...
+						reader = f
+						go func() {
+							<-logsEOFChan
+						}()
+					}
 				}
+				return jsonfile.Decode(os.Stdout, os.Stderr, reader, timestamps, since, until, logsEOFChan)
+			case "journald":
+				var journalctlArgs []string
+				journalctlArgs = append(journalctlArgs, fmt.Sprintf("SYSLOG_IDENTIFIER=%s", found.Container.ID()[:12]))
```
Comment (AkihiroSuda) [inline cmd/nerdctl/logs.go:172]: Please use a const or at least add a comment line for `12`

File: cmd/nerdctl/logs.go [hunk 12]
```diff
@@ -76,80 +78,113 @@ func logsAction(cmd *cobra.Command, args []string) error {
...
+				if follow {
+					journalctlArgs = append(journalctlArgs, "-f")
+				}
+				if since != "" {
+					journalctlArgs = append(journalctlArgs, "--since", since)
```
Comment (AkihiroSuda) [inline cmd/nerdctl/logs.go:177]:
```suggestion
					journalctlArgs = append(journalctlArgs, "--since", since + " ago")
```

File: cmd/nerdctl/logs.go [hunk 13]
```diff
@@ -76,80 +78,113 @@ func logsAction(cmd *cobra.Command, args []string) error {
...
+						defer f.Close()
+						reader = f
+						go func() {
+							<-logsEOFChan
+						}()
+					}
 				}
+				return jsonfile.Decode(os.Stdout, os.Stderr, reader, timestamps, since, until, logsEOFChan)
+			case "journald":
+				var journalctlArgs []string
```
Comment (AkihiroSuda) [inline cmd/nerdctl/logs.go:171]: Please add ` --output=cat` to eliminate extra header

File: pkg/logging/journald_logger.go [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+func (journaldLogger *JournaldLogger) Process(dataStore string, config *logging.Config) error {
+	if !journal.Enabled() {
+		return errors.New("the local systemd journal is not available for logging.")
+	}
+
+	var syslogIdentifier string
+	if _, ok := journaldLogger.Opts[Tag]; !ok {
+		syslogIdentifier = config.ID[:12]
```
Comment (AkihiroSuda) [inline pkg/logging/journald_logger.go:51]: Please add a const or at least add a comment line for this magic number

**Review by AkihiroSuda — APPROVED** [review]
Thanks
